#### Convex Optimization- Ex 6 ~ PROJECTED GRADIENT METHOD

__importing libraries to help us in minimizations__

In [1]:
import numpy as np
from numpy.linalg import norm
from scipy.optimize import minimize_scalar 
from scipy import optimize

In [2]:
#Calculation of gradient 
def gradf(x,y):
    theta_x = -400*y*x + 400*x**3 + 2*x - 2
    theta_y = 200*y - 200*x**2
    return np.array([theta_x, theta_y])

#calculation of f
def fx(myx):
    x=myx[0]
    y=myx[1]
    return 100*(y-x**2)**2 + (1-x)**2

def zeta(xk,gradf_xk,gamma):
    return xk - (1/gamma)*gradf(xk[0],xk[1])

def get_y(zeta_k):
    return zeta_k/(norm(zeta_k,2))

def delta(grad_xk,xk,yk):
    return np.dot(gradf(xk[0],xk[1]),yk-xk)

# minimization of step 4 , with bounded a between [0,1]
def to_min(a): 
    return fx(xk_global + a * (yk_global - xk_global)) - fx(xk_global)

gamma_list = [10**-1,10**-2,10**-3]# 10**-k for K =1,2,3

#tolerance init
tol = 10**-10
d=99999 # init delta with a big value for the first iteration

bounds_for_a = (0,1)

#### Implementing Projected Gradient Method

In [6]:
for i in gamma_list:
    k = 0
    gamma = i
    #doing these variables global so that
    #the to_min() function can see them.
    global xk_global
    global yk_global
    d=9999#something very big
    while abs(d) > tol:

        if k == 0: # first step

            #initialization of xk for k = 0
            xk = np.array([1/2, -1/2])
            zeta_k = zeta(xk,gradf(xk[0],xk[1]),gamma)
            yk    = get_y(zeta_k)
            d   = delta(gradf(xk[0],xk[1]),xk,yk)

            if abs(d) < tol:
                break
            else:
                xk_global = xk
                yk_global = yk
                ak = minimize_scalar(to_min,bounds = bounds_for_a ,method='bounded').x

            xk += ak*(yk-xk)
            k+=1
            #########################################################
        if k>0: #k=1,2,... until convergence
            zeta_k = zeta(xk,gradf(xk[0],xk[1]),gamma)
            yk = get_y(zeta_k)
            d = delta(gradf(xk[0],xk[1]),xk,yk)
            if abs(d) < tol:
                break
            else:
                xk_global = xk
                yk_global = yk
                ak = minimize_scalar(to_min,bounds = bounds_for_a ,method='bounded').x
            xk +=  ak * (yk-xk)

            k+=1
    
    print("Gamma Value: ",gamma)
    print("Number of iterations",k)
    print("Delta = %.10f"%(d))
    print("Final Solution",xk)
    if (xk[0]**2 + xk[1]**2<=1):
        print("solution acceptable")
    print("Value of function f on that point is: ",fx(xk))
    print("------------------------------------------------------")

Gamma Value:  0.1
Number of iterations 297
Delta = -0.0000000000
Final Solution [0.78641516 0.61769831]
solution acceptable
Value of function f on that point is:  0.04567480873434908
------------------------------------------------------
Gamma Value:  0.01
Number of iterations 489
Delta = -0.0000000001
Final Solution [0.78641515 0.61769831]
solution acceptable
Value of function f on that point is:  0.04567480876767872
------------------------------------------------------
Gamma Value:  0.001
Number of iterations 660
Delta = -0.0000000001
Final Solution [0.78641515 0.61769832]
solution acceptable
Value of function f on that point is:  0.045674808804133646
------------------------------------------------------


__We can observe that our algorithm converged really close to the real solution of the constrained optimization problem__

real solution is : x=0.7864, y= 0.6177 with F(x,y) = 0.0456775167 and 0.7864**2 + 0.6177**2 = 0.99997825<=1 

Reference:
https://www.mathworks.com/help/optim/ug/example-nonlinear-constrained-minimization.html

Difference of our solution between real 

In [4]:
0.04567480873434908 -  0.0456775167

-2.707965650917965e-06

In [9]:
0.04567480873795899 > 0.04567480873434908

True